### PR TITLE
[X86][AVX512] Check input-types to COMX

### LINKED ISSUE #118605: LLVM ERROR: Cannot select: ... = X86ISD::UCOMX ..., ...
Compiler Explorer: https://godbolt.org/z/GTfqbqGef

```
LLVM ERROR: Cannot select: t30: i32 = X86ISD::UCOMX t3, ConstantFP:f80<APFloat(0)>
  t3: f80,ch = load<(load (s80) from %fixed-stack.0, align 16)> t0, FrameIndex:i64<-1>, undef:i64
    t1: i64 = FrameIndex<-1>
    t2: i64 = undef
  t4: f80 = ConstantFP<APFloat(0)>
In function: ldequals
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /opt/compiler-explorer/clang-assertions-trunk/bin/llc -o /app/output.s -x86-asm-syntax=intel -mcpu=diamondrapids <source>
1.	Running pass 'Function Pass Manager' on module '<source>'.
2.	Running pass 'X86 DAG->DAG Instruction Selection' on function '@ldequals'
 #0 0x0000000003c0a0c8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x3c0a0c8)
 #1 0x0000000003c07abc SignalHandler(int) Signals.cpp:0:0
 #2 0x000076cacee42520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #3 0x000076cacee969fc pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0x969fc)
 #4 0x000076cacee42476 gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x42476)
 #5 0x000076cacee287f3 abort (/lib/x86_64-linux-gnu/libc.so.6+0x287f3)
 #6 0x0000000000758ec1 (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x758ec1)
 #7 0x00000000039b4078 llvm::SelectionDAGISel::CannotYetSelect(llvm::SDNode*) (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x39b4078)
 #8 0x00000000039b9519 llvm::SelectionDAGISel::SelectCodeCommon(llvm::SDNode*, unsigned char const*, unsigned int) (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x39b9519)
 #9 0x0000000001fac011 (anonymous namespace)::X86DAGToDAGISel::Select(llvm::SDNode*) X86ISelDAGToDAG.cpp:0:0
#10 0x00000000039b05eb llvm::SelectionDAGISel::DoInstructionSelection() (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x39b05eb)
#11 0x00000000039bf9aa llvm::SelectionDAGISel::CodeGenAndEmitDAG() (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x39bf9aa)
#12 0x00000000039c2f29 llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x39c2f29)
#13 0x00000000039c41e0 llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x39c41e0)
#14 0x00000000039b4a8f llvm::SelectionDAGISelLegacy::runOnMachineFunction(llvm::MachineFunction&) (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x39b4a8f)
#15 0x0000000002b70aa9 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (.part.0) MachineFunctionPass.cpp:0:0
#16 0x0000000003175f8f llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x3175f8f)
#17 0x0000000003176341 llvm::FPPassManager::runOnModule(llvm::Module&) (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x3176341)
#18 0x0000000003176be1 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x3176be1)
#19 0x000000000086e8f8 compileModule(char**, llvm::LLVMContext&) llc.cpp:0:0
#20 0x000000000076069e main (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x76069e)
#21 0x000076cacee29d90 (/lib/x86_64-linux-gnu/libc.so.6+0x29d90)
#22 0x000076cacee29e40 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e40)
#23 0x000000000086522e _start (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x86522e)
Program terminated with signal: SIGSEGV
Compiler returned: 139
```

### PR DESCRIPTION
**Pull request by abhishek-kaushik22**
Supported types for COMX are f16, f32 and f64. 
Without this check there's a crash on f80 types.
Fixes: https://github.com/llvm/llvm-project/issues/118605

### CODE REVIEW

**Comment by abhishek-kaushik22** [comment]
@mahesh-attarde @phoebewang @e-kud can you please review?

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 1]
```diff
@@ -24228,8 +24228,13 @@ SDValue X86TargetLowering::LowerSETCC(SDValue Op, SelectionDAG &DAG) const {
   if (Subtarget.hasAVX10_2()) {
     if (CC == ISD::SETOEQ || CC == ISD::SETUNE) {
       auto NewCC = (CC == ISD::SETOEQ) ? X86::COND_E : (X86::COND_NE);
-      return getSETCC(NewCC, DAG.getNode(X86ISD::UCOMX, dl, MVT::i32, Op0, Op1),
-                      dl, DAG);
+      auto isValidType = [&](MVT Type) {
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:24231]: `IsValidType`

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 2]
```diff
@@ -24228,8 +24228,13 @@ SDValue X86TargetLowering::LowerSETCC(SDValue Op, SelectionDAG &DAG) const {
...
+        return Type == MVT::f16 || Type == MVT::f32 || Type == MVT::f64;
+      };
+      if (isValidType(Op0.getSimpleValueType()) &&
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:24234]: I think checking single operand should be enough.

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 3]
```diff
@@ -24228,8 +24228,13 @@ SDValue X86TargetLowering::LowerSETCC(SDValue Op, SelectionDAG &DAG) const {
   if (Subtarget.hasAVX10_2()) {
     if (CC == ISD::SETOEQ || CC == ISD::SETUNE) {
       auto NewCC = (CC == ISD::SETOEQ) ? X86::COND_E : (X86::COND_NE);
-      return getSETCC(NewCC, DAG.getNode(X86ISD::UCOMX, dl, MVT::i32, Op0, Op1),
-                      dl, DAG);
+      auto isValidType = [&](MVT Type) {
+        return Type == MVT::f16 || Type == MVT::f32 || Type == MVT::f64;
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:24232]: What's the root cause for this? My understanding is #113567 only set these 3 types `Custom`, how can we reach with f80? Is f80 special? I checked bf16 doesn't crash even without this change.

Comment (abhishek-kaushik22) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:24232]: There are 3 checks above this one

`if (isSoftF16(Op0.getValueType(), Subtarget))`, 
`if (Op0.getValueType() == MVT::f128)`, 
`if (Op0.getSimpleValueType().isInteger())`

So, it seems `f80` is the only type that can reach it. I will simplify the check for just this type.

Comment (phoebewang) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:24232]: `isSoftF16` return false for bf16 under AVX10.2. Can you investigate it a bit more?

Comment (abhishek-kaushik22) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:24232]: I wasn't able to create an example with `bf16` that reaches this code, in my example I see `bfloat` bitcasted to a `i16` and then promoted to `f32` during legalization.
```
Legalizing node: t4: f16,ch = CopyFromReg t0, Register:f16 %1
Legalizing node: t28: i16 = bitcast t4
Legalizing node: t31: f32 = bf16_to_fp t28
```

Comment (phoebewang) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:24232]: I see, thanks! I assumed bf16 type is legal with AVX10.2 but it's not. Maybe adding an assert for bf16.

**Review by phoebewang — APPROVED** [review]
LGTM.

**Comment by abhishek-kaushik22** [comment]
@phoebewang can you please merge?

**Pull request merged**